File: 2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.dart
/**
 * Definition for singly-linked list.
 * class ListNode {
 *   int val;
 *   ListNode? next;
 *   ListNode([this.val = 0, this.next]);
 * }
 */
class Solution {
  ListNode? mergeNodes(ListNode? head) {
    ListNode? res = ListNode();
    ListNode? cur = res;
    int curSum = 0;
    head = head?.next;
    while(head!=null){
      if(head.val!=0){
        curSum += head.val;
      }else{
        cur?.next = ListNode(curSum);
        cur = cur?.next;
        curSum = 0;
      }
      head = head.next;
    }
    return res.next;
  }
}

File: 2299-merge-nodes-in-between-zeros/merge-nodes-in-between-zeros.rs
impl Solution {
    pub fn merge_nodes(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut current = &mut dummy;
        let mut node = head.unwrap().next;
        let mut current_sum = 0;
        while let Some(inner_node) = node {
            if inner_node.val != 0 {
                current_sum += inner_node.val;
            } else {
                current.next = Some(Box::new(ListNode::new(current_sum)));
                current = current.next.as_mut().unwrap();
                current_sum = 0;
            }
            node = inner_node.next;
        }
        dummy.next
    }
}
